File: 백준/Silver/24523. 내 뒤에 나와 다른 수/내 뒤에 나와 다른 수.py
n = int(input())

arr = list(map(int,input().split()))

l,r=0,1
ans = []
while l<r:
    while r<n and arr[l]==arr[r]:
        r+=1
    if r==n: break
    for _ in range(r-l):
        ans.append(r+1)
    l, r = r, r+1

while len(ans)<n:
    ans.append(-1)

print(*ans)

File: 백준/Silver/24523. 내 뒤에 나와 다른 수/내 뒤에 나와 다른 수.cc
#include <iostream>

int main() {
    int n;
    
    std::cin >> n;

    int arr[n];
    for (int i=0; i<n;i++) {
        std::cin >> arr[i];
    }

    int ans[n];
    int l = 0, r = 1;

    for (int i=0; i<n; i++) {
        ans[i] = -1;
    }
    
    while (l < r) {
        while (r < n && arr[l] == arr[r]){
            r++;
        }
        if (r == n) break;
        while (l < r) {
            ans[l++] = r+1;
        }
        l = r;
        r = r+1;
    }

    for (int i=0; i<n; i++) {
        std::cout << ans[i] << ' ';
    }
}